src/Client.ts

```diff
@@ -0,0 +1,1232 @@
+          }
+        }
+      }
+    }
+
+    this.gateway = new ShardManager(shardAmount, newIntents);
+    this.gateway.on("raw", (e) => {
+      switch (e.name) {
+        case "READY":
+          this.user = new PrivateUser(this, e.data.user);
+          this.emit("shardReady", e.data.shard!.join("/"));
+          break;
+        case "RESUMED":
+          break;
+        case "RECONNECT":
+          break;
+
+        case "GUILD_CREATE": {
+          const guild = new GatewayGuild(this, e.data);
+          for (const channel of e.data.channels) {
+            this.guildChannels.set(
+              channel.id,
+              this.newChannelSwitch(channel) as TextBasedGuildChannels,
+            );
+          }
+          for (const { user } of e.data.members) {
+            this.users.set(user.id, new User(this, user));
+          }
+          this.guilds.set(e.data.id, guild);
+          this.emit("guildCreate", guild);
+          break;
+        }
+        case "GUILD_UPDATE": {
+          const oldGuild = this.guilds.get(e.data.id)!;
+          const newGuild = new GatewayGuild(this, e.data);
+          this.guilds.set(e.data.id, newGuild);
+          this.emit("guildUpdate", oldGuild, newGuild);
+          break;
+        }
+        case "GUILD_DELETE":
+          this.guilds.delete(e.data.id);
+          this.emit("guildDelete", e.data);
+          break;
+
+        case "GUILD_ROLE_CREATE": {
+          const guild = this.guilds.get(e.data.guild_id)!;
+          const role = new Role(this, e.data.role, e.data.guild_id);
+          guild.roles.set(e.data.role.id, role);
+          this.guilds.set(e.data.guild_id, guild);
+          this.emit("guildRoleCreate", guild, role);
+          break;
+        }
+        case "GUILD_ROLE_UPDATE": {
+          const guild = this.guilds.get(e.data.guild_id)!;
+          const oldRole = guild.roles.get(e.data.role.id)!;
+          const newRole = new Role(this, e.data.role, e.data.guild_id);
+          guild.roles.set(e.data.role.id, newRole);
+          this.guilds.set(e.data.guild_id, guild);
+          this.emit("guildRoleUpdate", guild, oldRole, newRole);
+          break;
+        }
+        case "GUILD_ROLE_DELETE": {
+          const guild = this.guilds.get(e.data.guild_id)!;
+          const role = guild.roles.get(e.data.role_id)!;
+          guild.roles.delete(e.data.role_id);
+          this.guilds.set(e.data.guild_id, guild);
+          this.emit("guildRoleDelete", guild, role);
+          break;
+        }
+
+        case "CHANNEL_CREATE": {
+          const channel = this.newChannelSwitch(e.data);
+          if (channel.type === "dm" || channel.type === "groupDM") {
+            this.dmChannels.set(channel.id, channel as DMChannels);
+          } else {
+            this.guildChannels.set(channel.id, channel as GuildChannels);
+
+            const guild = this.guilds.get((channel as GuildChannels).guildId)!;
+            guild.channels.set(channel.id, channel as GuildChannels);
```

Comment:
This sort of casting is error-prone. Consider creating a type guard, something like: 
```ts
function isDmChannel (channel: DMChannels | GuildChannels): channel is DMChannels {
  return channel.type === "dm" || channel.type === "groupDM"
}```
Although it should probably be able to infer the type correctly if each channel type is correctly specified, and the input type is a union of the two — which would avoid having to do a type guard function.

Reply:
You're doing more similar checks and casts further down, same applies to them, and this type guard function could be used on them too so you don't have to repeat yourself and give undue room for mistakes.